src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientServices.cs

```diff
@@ -24,32 +25,217 @@ namespace Microsoft.Azure.IIoT.OpcUa.Protocol.Services {
+            }
+
+            // Set the Certificate as the newly created certificate
+            Certificate = ownCertificate;
+
+            // updatecertificates validator
+            _OpcApplicationConfig.CertificateValidator.CertificateValidation += 
+                new Opc.Ua.CertificateValidationEventHandler(CertificateValidator_CertificateValidation);
+
+            await _OpcApplicationConfig.CertificateValidator.Update(_OpcApplicationConfig).ConfigureAwait(false);
+
+            if (_OpcApplicationConfig.SecurityConfiguration.AutoAcceptUntrustedCertificates) {
+                _logger.Warning("WARNING: Automatically accepting certificates. This is a security risk.");
+            }
+        }
+
+        /// <summary>
+        /// Event handler to validate certificates.
+        /// </summary>
+        private void CertificateValidator_CertificateValidation(
+            Opc.Ua.CertificateValidator validator, 
+            Opc.Ua.CertificateValidationEventArgs e){
+
+            if (e.Error.StatusCode == Opc.Ua.StatusCodes.BadCertificateUntrusted){
+
+                e.Accept = _configuration.AutoAcceptUntrustedCertificates;
+
+                if (_configuration.AutoAcceptUntrustedCertificates) {
+
+                    _logger.Information($"Certificate '{e.Certificate.Subject}' will be trusted, " +
+                        $"because AutoAcceptUntrustedCertificates is on.");
+
+                    try {
+
+                        ICertificateStore trustedStore = 
+                            _OpcApplicationConfig.SecurityConfiguration.TrustedPeerCertificates.OpenStore();
+                         try {
+                            _logger.Information($"Adding server certificate to trusted peer store. " +
+                                $"StorePath=" +
+                                $"{_OpcApplicationConfig.SecurityConfiguration.TrustedPeerCertificates.StorePath}");
+                            trustedStore.Delete(e.Certificate.Thumbprint);
+                            
+                            trustedStore.Add(e.Certificate);
+                            validator.Update(_OpcApplicationConfig.SecurityConfiguration);
+                        }
+                        catch (Exception ex) {
+                            _logger.Warning(ex, $"Can not add server certificate to " +
+                                $"trusted peer store.");
+                        }
+                        finally {
+                            trustedStore.Close();
+                        }
+                    }
+                    catch(Exception ex) {
+                        _logger.Warning(ex, $"Can not open certificate to trusted certificate peer store.");
+                    }
+                }                                
+            }
+            else{
+                _logger.Information($"Not trusting OPC application with the certificate subject '{e.Certificate.Subject}'.");
```

Comment:
1 log is better or else you get each line interspersed with other log outpu.

Reply:
done